lib/OpenQA/Schema/Result/Jobs.pm

```diff
@@ -1904,6 +1904,20 @@ sub investigate {
             $inv{needles_log} ||= 'No needle changes recorded, test regression due to needles unlikely';
             $inv{needles_diff_stat} = $self->git_diff($dir, $refspec_needles_range) if $inv{needles_log};
         }
+
+        my $worker_packages_file = $self->result_dir() . "/worker_packages.txt";
+        if (-e $worker_packages_file && -e $prev->result_dir() . "/worker_packages.txt") {
+            my @files_packages = map { Mojo::File->new($_->result_dir(), 'worker_packages.txt')->slurp } ($prev, $self);
+            my $diff_packages  = eval { diff(\$files_packages[0], \$files_packages[1], {CONTEXT => 0}) };
+            $inv{diff_packages_to_last_good} = join("\n", grep { !/(^@@|$ignore)/ } split(/\n/, $diff_packages));
+        }
+        elsif (-e $worker_packages_file) {
+            $inv{diff_packages_to_last_good} = "The list of packages for the last good is not available";
+        }
+        else {
+            $inv{diff_packages_to_last_good} = "The list of packages is not available";
+        }
```

Comment:
I would avoid the repeated filesystem lookup for `$worker_packages_file` and also the other repetitions. Also prefer single quotes if no variable substitutions are used.

```suggestion
        my $current_file = path($self->result_dir, 'worker_packages.txt');
        if (-e $current_file) {
            my $prev_file = path($prev->result_dir, 'worker_packages.txt');
            if (-e $prev_file) {
                my @files_packages = map { $_->slurp } ($prev_file, $current_file);
                my $diff_packages  = eval { diff(\$files_packages[0], \$files_packages[1], {CONTEXT => 0}) };
                $inv{diff_packages_to_last_good} = join("\n", grep { !/(^@@|$ignore)/ } split(/\n/, $diff_packages));
            } else {
                $inv{diff_packages_to_last_good} = 'The list of packages for the last good is not available';            
            }
        }
        else {
            $inv{diff_packages_to_last_good} = 'The list of packages is not available';
        }
```

It might also make sense to catch exceptions from `slurp`.

Reply:
I not sure about this part
```try {
                    my @files_packages = map { $_->slurp } ($prev_file, $current_file);
                    my $diff_packages  = eval { diff(\$files_packages[0], \$files_packages[1], {CONTEXT => 0}) };
                    $inv{diff_packages_to_last_good} = join("\n", grep { !/(^@@|$ignore)/ } split(/\n/, $diff_packages));
                } catch {
                    $inv{diff_packages_to_last_good} = 'An problem occurred getting the packages file content. And couldn\'t be compared';

Reply:
actually I would do it differently. As so far I think the method `investigate` is rather clean and works without any additional blank lines for separation that asks for an extracted method. Then you can just do something like:

```perl
sub packages_diff {
    my ($self, $prev) = @_;
    my $current_file = path($self->result_dir, 'worker_packages.txt');
    return unless -e $current_file;
    my $prev_file = path($prev->result_dir, 'worker_packages.txt');
    return unless -e $prev_file;
    my @files_packages = map { $_->slurp } ($prev_file, $current_file);
    my $diff_packages  = eval { diff(\$files_packages[0], \$files_packages[1], {CONTEXT => 0}) };
    return join("\n", grep { !/(^@@|$ignore)/ } split(/\n/, $diff_packages));
}

…

$inv{diff_packages_to_last_good} = $self->packages_diff($prev) // 'Diff of packages not available';
```

This way we have the same early exits in case any file does not exist, no multiple filesystem operations, no additional indendation and even possibility for easier mocking of methods to test like in this case `packages_diff`

Reply:
Ok, I put this commit as a new commit